### PR TITLE
perf(auth): optimize auth token handling #15763

### PR DESCRIPTION
**Pull request by Sahil-4555**
**What type of PR is this?**
> Other

**What does this PR do? Why is it needed?**
This PR applies micro-optimizations to the auth token handling code. It improves efficiency and readability by reducing unnecessary allocations and adding an early length check before performing constant-time comparison. [PR#15763](https://github.com/OffchainLabs/prysm/pull/15763)

**Changes Included**
- Use `strings.HasPrefix` + slicing instead of `strings.Split` to avoid allocations
- Add early length check before `subtle.ConstantTimeCompare`

**Which issues(s) does this PR fix?**
No functional changes or security fixes; this PR improves performance and code clarity in the auth token handling logic.

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

### CODE REVIEW

File: validator/rpc/intercepter.go [hunk 1]
```diff
@@ -68,15 +68,13 @@ func (s *Server) authorize(ctx context.Context) error {
 		return status.Errorf(codes.InvalidArgument, "Retrieving metadata failed")
 	}
 
-	authHeader, ok := md["authorization"]
-	if !ok {
-		return status.Errorf(codes.Unauthenticated, "Authorization token could not be found")
-	}
-	if len(authHeader) < 1 || !strings.Contains(authHeader[0], "Bearer ") {
+	authHeader := md["authorization"][0]
```
Comment (james-prysm) [inline validator/rpc/intercepter.go]: can this panic? the old one checks authHeader, ok := md["authorization"]

Comment (Sahil-4555) [reply validator/rpc/intercepter.go]: Yes, you’re right. This can panic if the authorization metadata is missing or empty. The original code handled this safely by checking both conditions first. I’ve added that guard back before indexing [0] and reverted to the original safe behavior.

File: validator/rpc/intercepter.go [hunk 2]
```diff
@@ -45,14 +48,13 @@ func (s *Server) AuthTokenHandler(next http.Handler) http.Handler {
 				httputil.HandleError(w, "Unauthorized: no Authorization header passed. Please use an Authorization header with the jwt created in the prysm wallet", http.StatusUnauthorized)
 				return
 			}
-			tokenParts := strings.Split(reqToken, "Bearer ")
-			if len(tokenParts) != 2 {
+			if !strings.HasPrefix(reqToken, bearerPrefix) {
```
Comment (james-prysm) [inline validator/rpc/intercepter.go]: token, ok := strings.CutPrefix(reqToken, bearerPrefix)                   
  if !ok {                                                                 
      // invalid format                                                    
  }      
  
  what about usign cut prefix?

**Comment by james-prysm** [comment]
I think with such a change we should add some unit tests, would be nice with a bench mark too

1. Add test for malformed Bearer prefix: "Bearertoken" (missing space)   
  should return StatusBadRequest                                           
  2. Add test for empty token after Bearer: "Bearer " should return        
  appropriate error    
  
  if this is just for readability it's a lower priority

**Comment by james-prysm** [comment]
if this covers more cases than the original it would be good to add tests for that, it was written quite a long time ago

**Comment by Sahil-4555** [comment]
```
goos: linux
goarch: amd64
pkg: github.com/OffchainLabs/prysm/v7/validator/rpc
cpu: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
                                    │ old_bench.txt │            new_bench.txt            │
                                    │    sec/op     │   sec/op     vs base                │
AuthTokenHandler_ValidToken-8          481.9n ± 23%   356.6n ± 1%  -26.00% (p=0.000 n=10)
AuthTokenHandler_InvalidToken-8        2.079µ ± 12%   1.729µ ± 9%  -16.84% (p=0.000 n=10)
AuthTokenInterceptor_ValidToken-8      2.084µ ±  3%   1.637µ ± 4%  -21.47% (p=0.000 n=10)
AuthTokenInterceptor_InvalidToken-8    927.7n ± 14%   637.0n ± 8%  -31.33% (p=0.000 n=10)
geomean                                1.180µ         895.4n       -24.10%

                                    │ old_bench.txt │            new_bench.txt             │
                                    │     B/op      │     B/op      vs base                │
AuthTokenHandler_ValidToken-8            240.0 ± 0%     208.0 ± 0%  -13.33% (p=0.000 n=10)
AuthTokenHandler_InvalidToken-8        1.195Ki ± 0%   1.164Ki ± 0%   -2.61% (p=0.000 n=10)
AuthTokenInterceptor_ValidToken-8      1.328Ki ± 0%   1.297Ki ± 0%   -2.35% (p=0.000 n=10)
AuthTokenInterceptor_InvalidToken-8      608.0 ± 0%     528.0 ± 0%  -13.16% (p=0.000 n=10)
geomean                                  702.0          645.7        -8.02%

                                    │ old_bench.txt │           new_bench.txt            │
                                    │   allocs/op   │ allocs/op   vs base                │
AuthTokenHandler_ValidToken-8            5.000 ± 0%   4.000 ± 0%  -20.00% (p=0.000 n=10)
AuthTokenHandler_InvalidToken-8          15.00 ± 0%   14.00 ± 0%   -6.67% (p=0.000 n=10)
AuthTokenInterceptor_ValidToken-8        11.00 ± 0%   10.00 ± 0%   -9.09% (p=0.000 n=10)
AuthTokenInterceptor_InvalidToken-8      8.000 ± 0%   6.000 ± 0%  -25.00% (p=0.000 n=10)
geomean                                  9.013        7.614       -15.53%
```

**Comment by Sahil-4555** [comment]
> I think with such a change we should add some unit tests, would be nice with a bench mark too
> 
> 1. Add test for malformed Bearer prefix: "Bearertoken" (missing space)
>    should return StatusBadRequest
> 2. Add test for empty token after Bearer: "Bearer " should return
>    appropriate error
> 
> if this is just for readability it's a lower priority

Thanks for the suggestion.
I’ve tried to cover most of the points mentioned in your comment.
I added unit tests for the malformed Bearer prefix (Bearertoken) and for the empty token case (Bearer ), and updated the code for better readability using CutPrefix.
I also added benchmarks, though I understand they are lower priority.

Please let me know if you’d like me to remove or add anything, or adjust the scope further.

**Comment by Sahil-4555** [comment]
Hi, @james-prysm any chance of getting this added?

**Review by james-prysm — DISMISSED** [review]
thanks, will include now with better unit tests

File: validator/rpc/intercepter.go (hunk 2 above)
Comment (Sahil-4555) [reply validator/rpc/intercepter.go]: Done.

File: validator/rpc/intercepter_test.go [hunk 3]
```diff
@@ -275,93 +273,39 @@ func BenchmarkServer_AuthTokenHandler(b *testing.B) {
...
-					b.Fatalf("unexpected error: %v", err)
-				}
-			}
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, err := interceptor(ctx, "xyz", unaryInfo, func(ctx context.Context, req any) (any, error) {
+			return nil, nil
 		})
+		if err != nil {
```
Comment (james-prysm) [inline validator/rpc/intercepter_test.go]: why not add a require no error here instead of b.Fatal

Comment (Sahil-4555) [reply validator/rpc/intercepter_test.go]: Done.

**Comment by james-prysm** [comment]
seems like you have linting issues

**Comment by Sahil-4555** [comment]
> seems like you have linting issues

Fixed

**Review by james-prysm — APPROVED** [review]
Thanks for the contribution
